### ISSUE TITLE
HISTCONTROL ignorespace

### ISSUE THREAD

**Issue by nh2** [xonsh/xonsh#2418]
Many shells [have a feature](
https://unix.stackexchange.com/questions/115917/why-is-bash-not-storing-commands-that-start-with-spaces) that you can set `HISTCONTROL` to `ignorespace` to prevent commands starting with a space from going into history.

That is so that you can conveniently set an environment variable containing secret information that shouldn't go into history without having to resort to something like `getpass()`, or so that you can run a command that you don't want in the history to prevent yourself from accidentally running it again later, such as ` rm -rf *`.

**Comment by nh2** [comment]
I took a quick look at what would be needed for this, and the logic change is trivial, like:

```diff
diff --git a/xonsh/history/json.py b/xonsh/history/json.py
index 9c11babe..d6eca11d 100644
--- a/xonsh/history/json.py
+++ b/xonsh/history/json.py
@@ -201,6 +201,9 @@ class JsonHistoryFlusher(threading.Thread):
             if 'ignoreerr' in opts and cmd['rtn'] != 0:
                 # Skipping failed cmd
                 continue
+            if 'ignorespace' in opts and cmd['inp'].startswith(' '):
+                # Skipping cmd starting with space
+                continue
             cmds.append(cmd)
             last_inp = cmd['inp']
         with open(self.filename, 'r', newline='\n') as f:
diff --git a/xonsh/history/sqlite.py b/xonsh/history/sqlite.py
index 0446699e..4648d4da 100644
--- a/xonsh/history/sqlite.py
+++ b/xonsh/history/sqlite.py
@@ -202,6 +202,12 @@ class SqliteHistory(History):
         if 'ignoreerr' in opts and cmd['rtn'] != 0:
             # Skipping failed cmd
             return
+        print(type(cmd))
+        print(cmd)
+        raise 'here'
+        if 'ignorespace' in opts and cmd['inp'].startswith(' '):
+            # Skipping cmd starting with space
+            return
```

But the problem is that `cmd['inp']` for some reason already has leading whitespace stripped.

It seems that `xonsh/ptk/shell.py`'s `line = self.singleline(auto_suggest=auto_suggest)` already returns it stripped.

I haven't looked deeper yet.

**Comment by hugo-aa** [comment]
Hi @nh2, thanks for your inputs here. I think this is a nice feature.

> the problem is that cmd['inp'] for some reason already has leading whitespace stripped.

There is no particular reason with this - it, I guess, strips the inputs so that item saved in history is cleaner.

Could you create an PR for this, so that we can review it, to make this feature added?

**Comment by mitnk** [comment]
Sorry, forgot to switch my account :( use the correct account now.

Hi @nh2, thanks for your inputs here. I think this is a nice feature.

> the problem is that cmd['inp'] for some reason already has leading whitespace stripped.

There is no particular reason with this - it, I guess, strips the inputs so that item saved in history is cleaner.

Could you create an PR for this, so that we can review it, to make this feature added?

**Comment by nh2** [comment]
> Could you create an PR for this, so that we can review it, to make this feature added?

Would like to but I haven't found yet where that witespace stripping happens.

**Comment by mitnk** [comment]
Check a bit - the input is stripped in `xonsh.base_shell::BaseShell.precmd()`. But updating it would cause xonsh raise compile error. So yes, it seems not an easy change - Needs either change xonsh's parser, or use a xonsh context variable to store this `prefixed-by-a-space` info.

There may be other ways that I don't know?

**Comment by astronouth7303** [comment]
Does the whitespace stripping happen before or after the preprocessing stage?

**Comment by nh2** [comment]
@astronouth7303 Not sure if that answers your question, but it is literally the `precmd()` that @mitnk mentioned that does the `lstrip()`.

The problem is that if I remove the `lstrip()`, xonsh's parser rejects the line, and I'm not very familiar with Python YACC.

**Comment by laerus** [comment]
I think this was raised in the past too, agreed with @mitnk this would require a change in the parser which atm ignores white space like python!

we are trying to contain bash-like features in the bashism xontrib, but we might need some changes in xonsh in order for this to be implemented, with more hooks events could handle this aswell.

**Comment by astronouth7303** [comment]
I think the only change that needs to be made in core xonsh is to make sure that:

1. The original command (as the user typed it) is stored and passed along
2. The stripping is done at some later stage, so the "original command" includes any whitespace.

Everything else is policy, and arguments can be made for making it a core option, requiring a proxy history backend, or some other kind of xontrib hook.

**Comment by nh2** [comment]
@astronouth7303 This makes sense to me; it seems that it's stripping just before feeding it into the parser would be good enough.

I had a short shot at implementing that but didn't succed because of multiline input that somehow made the issue more complex, but that might just be due to unfamiliarity.

**Comment by laerus** [comment]
multiline is probably why the parser needs to change,
imho the feature doesnt worth the complexity it requires on the shell loop,
even if we change the parser it would be in a way that would benefit xonsh in general
and not just the history part.

**Comment by gforsyth** [comment]
Closed via #4062
